### PR TITLE
LKMM Dep Checker Alpha

### PR DESCRIPTION
**Pull request by PBHDK**
## Running the passes

Build clang as per usual. The passes are enabled by passing `fsanitize=lkmm-dep-checker` to the clang driver.

For building Linux, e.g. with defconfig, the following arguments are required for `make`:
`"CC=<path_to_clang>" "ARCH=arm64" "CROSS_COMPILE=aarch64-unknown-linux-gnu-"  "KCFLAGS=-g" "KCFLAGS=-fsanitize=lkmm-dep-checker"`.

### CODE REVIEW

File: llvm/include/llvm/Transforms/Utils/LKMMDependenceAnalysis.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+#include <queue>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+#ifndef LLVM_TRANSFORMS_UTILS_LKMMDEPENDENCEANALYSIS_H
+#define LLVM_TRANSFORMS_UTILS_LKMMDEPENDENCEANALYSIS_H
+
+namespace llvm {
+namespace {
```
Comment (melver) [inline llvm/include/llvm/Transforms/Utils/LKMMDependenceAnalysis.h:30]: anonymous namespaces in headers don't make sense.

move it inside the class.

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 2]
```diff
@@ -1413,6 +1414,24 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
             FPM.addPass(BoundsCheckingPass());
           });
 
+    if (LangOpts.Sanitize.has(SanitizerKind::LKMMDepChecker)) {
+      if (CodeGenOpts.OptimizationLevel == 0) {
+        PB.registerOptimizerLastEPCallback(
+            [](ModulePassManager &MPM, OptimizationLevel Level) {
+              MPM.addPass(AnnotateLKMMDeps());
```
Comment (melver) [inline clang/lib/CodeGen/BackendUtil.cpp:1421]: For debugging?

File: llvm/lib/Passes/PassRegistry.def [hunk 3]
```diff
@@ -119,6 +119,8 @@ MODULE_PASS("sancov-module", ModuleSanitizerCoveragePass())
 MODULE_PASS("memprof-module", ModuleMemProfilerPass())
 MODULE_PASS("poison-checking", PoisonCheckingPass())
 MODULE_PASS("pseudo-probe-update", PseudoProbeUpdatePass())
+MODULE_PASS("annotate-deps", AnnotateLKMMDeps())
```
Comment (melver) [inline llvm/lib/Passes/PassRegistry.def:122]: s/annotate-deps/lkmm-annotate-deps/

File: llvm/lib/Passes/PassRegistry.def [hunk 4]
```diff
@@ -119,6 +119,8 @@ MODULE_PASS("sancov-module", ModuleSanitizerCoveragePass())
...
+MODULE_PASS("verify-deps", VerifyLKMMDeps())
```
Comment (melver) [inline llvm/lib/Passes/PassRegistry.def:123]: s/verify-deps/lkmm-verify-deps/

File: llvm/lib/Transforms/Utils/LKMMDependenceAnalysis.cpp [hunk 5]
```diff
@@ -0,0 +1,2011 @@
...
+#include "llvm/IR/Metadata.h"
+#include "llvm/IR/Module.h"
+
+// TODO add clang support for clang -fsanitize=lkmm
+// TODO strict and relaxed mode
+
+namespace llvm {
+namespace {
+
+#define INTERPROCEDURAL_REC_LIMIT_ANNOTATION 3
```
Comment (melver) [inline llvm/lib/Transforms/Utils/LKMMDependenceAnalysis.cpp:37]: In C++ avoid constants via '#define'. 

https://abseil.io/tips/140

So probably 'constexpr int InterproceduralRecLimitAnnotation = 3;
...

File: llvm/lib/Transforms/Utils/LKMMDependenceAnalysis.cpp [hunk 6]
```diff
@@ -0,0 +1,2011 @@
...
+
+void PotAddrDepBeg::addAdrDep(std::string PathTo2, Instruction *I2,
+                              bool FDep) const {
+  if (beginsAt(I2))
+    return;
+
+  auto ID = getID() + PathTo2;
+
+  std::string begin_annotation =
+      "DoitLk: address dep begin," + ID + "," + getPathTo() + ";";
```
Comment (melver) [inline llvm/lib/Transforms/Utils/LKMMDependenceAnalysis.cpp:1138]: "DoitLk"? Maybe no longer use the "codename". ;-)

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 2 above)
Comment (PBHDK) [reply clang/lib/CodeGen/BackendUtil.cpp]: Yup. But I just saw there's another workaround: `clang -O1 -mllvm -disable-llvm-optzns -disable-llvm-passes`. If I use that instead, I can get rid of the above.

**Pull request merged**